impf/decorators.py

```diff
@@ -65,7 +66,7 @@ def func(self, *args, **kwargs):
             if self.keep_browser:
                 self.driver.close()
                 self.__post_init__()
-                return self.control_main()
+            raise WorkflowException('Got StaleElementReferenceException!')
```

Comment:
I think StaleElements usually occurr when the bot detatches from the browser. Reinitializing would be desirable here to reattach

Reply:
If you reinitialize, then the browser will be open with an empty tab. Calling `control_main` will result in an AssertionError in `location_page` if `error_counter != 0`.
But even otherwise `return self.control_main()` would not solve the issue what happens to the control flow when that call returns.

An example:
main -> control_main -> location_page -> StaleElements -> control_errors -> control_main -> "all works fine, but no vacancies" -> return from control_main -> return from control_errors -> "now the remaining part of the first control_main gets executed, which will be either control_vermittlungscode or confirm_eligible as has_vacancy will ignore its TimeoutException" -> TimeoutException or AssertionError from control_vermittlungscode/confirm_eligible -> control_errors -> return control_main -> "and now everything repeats"

Reply:
Why would the rest of `control_main` be executed – if an exception occurs in `control_main`, resulting in the try-catch block of the decorator, no further execution in `control_main` is done afterwards, as the function exited with an exception.


The second part is true:

> confirm_eligible as has_vacancy will ignore its TimeoutException" -> TimeoutException or AssertionError from control_vermittlungscode/confirm_eligible -> control_errors -> return control_main -> "and now everything repeats"

But `control_main` should give us a fresh start if an `AssertionError` occurs.

Reply:
Sorry, I missed that you were waiting for feedback

> Why would the rest of `control_main` be executed – if an exception occurs in `control_main`, resulting in the try-catch block of the decorator, no further execution in `control_main` is done afterwards, as the function exited with an exception.

There are several instances of `control_main` called recursively, I will try to make that clearer in my example:

main -> control_main(1) -> location_page -> StaleElements -> control_errors -> control_main(2) -> "all works fine, but no vacancies" -> return from control_main(2) -> return from control_errors -> "now the remaining part of the first control_main(1) gets executed, which will be either control_vermittlungscode or confirm_eligible as has_vacancy will ignore its TimeoutException" -> TimeoutException or AssertionError from control_vermittlungscode/confirm_eligible -> control_errors -> control_main(3) -> "and now everything repeats"

`control_main(1)` calls `location_page` where some error happens which is caught and a second instance of `control_main(2)` is called. When that second `control_main(2)` returns (normally), there is still `control_main(1)` being executed and that will now continue to do things.

I hope that makes it a bit clearer.

Cheers,
Volker

Reply:
Hey @lusor, thanks for the feedback. I think it's slowly starting to make sense. I have also experienced the recursion issue you mentioned recently, so it's definitely a problem – but

I still don't 100% get it yet tho – if `control_main(1)` is called and `location_page` throws an exception, the exception is escalated to `control_main` -> `control_errors`.  

Because `control_errors` is a decorator, `control_main` raising an exception means `control_main` has stopped its execution and exited raising an exception.

Where am I going wrong here?

Thanks, 
Alfons

Reply:
Damn, I just realized that you're right:

> Because `control_errors` is a decorator, `control_main` raising an exception means `control_main` has stopped its execution and exited raising an exception.

`control_main(1)` really has stopped execution, so it won't continue. However, there are (more complex) situations (like with failed shadow ban, I think), where the code starts another `control_main` while not in `control_errors` and then the recursion starts.

As I'm not a developer, I'm not sure, what the right/correct/better architecture would be. The `WorkflowException` I introduced is a bit hacky because I didn't (and still don't) want to completely rewrite the whole code.
I guess, some kind of state machine could be a solution here. With that there would be no need for nested method calls and potential recursion. Recovery would be quite simple, too, as whatever happens the machine only needs to figure out what state the browser is in and act accordingly. But again, this would be a major rewrite of the code.

Reply:
Glad we're on the same page now. You're right – a major code rewrite would be required, since we're stacking a lot of decorator- and return calls. Didn't consider that when originally creating the bot. I'm not a developer either. ;)

As the error only really occurs after the bot has run for some time, I think the most practical and easiest fix would be to just `pass` every `RecursionError` in `main.py.impf_me`, allowing the bot to simply restart the whole procedure whenever the call stack gets too big.

After all the bot is about being reliable and usually runs for 1-2 days, so we don't need the *perfect* solution geared towards running at the theoretical max efficiency. What do you think?

BR, Alfons

Reply:
I'm not sure, if I'm handling some more cases with this PR, but yes, catching `RecursionError` in `impf_me` would fix the most serious issue and would be quite simple. So if you're not comfortable with my more complex changes , make it so!

Cheers,
Volker

Reply:
Can – but I'll also leave you the honors and credit after originally bringing it up, in case you want to create a PR.

Otherwise I can also do it real quick – no worries.

BR, Alfons

Reply:
Just do it :)

Cheers,
Volker